src/main/scala/org/graphframes/lib/TriangleCount.scala

```diff
@@ -50,15 +42,46 @@ class TriangleCount private[graphframes] (private val graph: GraphFrame) extends
   }
 }
 
-// TODO(tjh) graphX requires the graph to be partitioned.
 private object TriangleCount {
 
   private def run(graph: GraphFrame): GraphFrame = {
-    val gx = graphxlib.TriangleCount.run(graph.cachedTopologyGraphX)
-    GraphXConversions.fromGraphX(graph, gx, Seq(COUNT_ID), Nil)
-  }
+    // Dedup edges by flipping them to have LONG_SRC < LONG_DST
+    val e = graph.indexedEdges
+    val min = udf { (x: Long, y: Long) => math.min(x, y) }
+    val max = udf { (x: Long, y: Long) => math.max(x, y) }
+    val dedupedE = e.select(e(SRC), e(DST),
+      min(e(LONG_SRC), e(LONG_DST)).as(LONG_SRC),
+      max(e(LONG_SRC), e(LONG_DST)).as(LONG_DST))
+      .dropDuplicates(Seq(LONG_SRC, LONG_DST))
+      .select(SRC, DST)
+    val g2 = GraphFrame(graph.vertices, dedupedE)
 
-  private val COUNT_ID = "count"
+    // 2 types of triangles:
+    // - cycles: edges forming a cycle.  These are counted 3 times each by motif finding.
+    // - noncycles: cycle with one edge flipped.  These are counted 1 time each by motif finding.
+    val cycles = g2.find("(a)-[]->(b); (b)-[]->(c); (c)-[]->(a)")
+      .select(col("a.id").as("a"), col("b.id").as("b"), col("c.id").as("c"))
+      .filter("a != b AND b != c AND c != a")
+    val noncycles = g2.find("(a)-[]->(b); (b)-[]->(c); (a)-[]->(c)")
+      .select(col("a.id").as("a"), col("b.id").as("b"), col("c.id").as("c"))
+      .filter("a != b AND b != c AND c != a")
 
+    // TODO: Can we make this more efficient using lateral view explode?
+    // Count the triangle for each vertex.
+    // - Since cycles are counted 3x, we can just keep "a"
+    // - Since noncycles are counted once, we need to count each of a,b,c
+    val allTriangles = noncycles.select(col("a").as(ID))
+      .unionAll(noncycles.select(col("b").as(ID)))
+      .unionAll(noncycles.select(col("c").as(ID)))
+      .unionAll(cycles.select(col("a").as(ID)))
+    val triangleCounts = allTriangles.groupBy(ID).count()
+    val v = graph.vertices
+    val countsCol = when(triangleCounts("count").isNotNull, triangleCounts("count"))
+      .otherwise(0).as(COUNT_ID)
+    val newV = v.join(triangleCounts, v(ID) === triangleCounts(ID), "left_outer")
+      .select(countsCol +: v.columns.map(v.apply) :_ *)
```

Comment:
I believe the only columns in "triangleCounts" are "id" and "count" but I'll add a test.